File: packages/python/src/slybrowser/automation.py
"""Automation backend selection and framework binding compatibility."""

from __future__ import annotations

from importlib import metadata
from typing import Literal, TypedDict

from .errors import ConfigurationError

AutomationBackend = Literal["project-webdriver", "playwright"]


class AutomationCapability(TypedDict, total=False):
    backend: AutomationBackend
    language: Literal["python"]
    framework_version: str
    native_humanize: bool
    persistent_context: bool


_SUPPORTED_PLAYWRIGHT_LINES = frozenset({"1.62"})


def validate_playwright_version(version: str) -> str:
    parts = version.split("+", 1)[0].split("-", 1)[0].split(".")
    if len(parts) < 2 or not all(part.isdigit() for part in parts):
        raise ConfigurationError(
            f"Invalid Playwright version: {version}",
            code="framework_version_invalid",
        )
    line = ".".join(parts[:2])
    if line not in _SUPPORTED_PLAYWRIGHT_LINES:
        supported = ", ".join(sorted(_SUPPORTED_PLAYWRIGHT_LINES))
        raise ConfigurationError(
            f"Unsupported Playwright version {version}; supported lines: {supported}",
            code="framework_version_unsupported",
        )
    return version


def resolve_playwright_version(explicit: str | None = None) -> str:
    if explicit is not None:
        return validate_playwright_version(explicit)
    try:
        return validate_playwright_version(metadata.version("playwright"))
    except metadata.PackageNotFoundError as error:
        raise ConfigurationError(
            "Unable to determine the installed Playwright package version; "
            "pass framework_version explicitly",
            code="framework_version_missing",
        ) from error


def require_playwright_humanize_support(requested: bool) -> None:
    _ = requested


def automation_capability(
    backend: AutomationBackend = "project-webdriver",
    *,
    framework_version: str | None = None,
) -> AutomationCapability:
    if backend == "project-webdriver":
        if framework_version is not None:
            raise ConfigurationError(
                "Project WebDriver does not accept framework_version",
                code="framework_version_forbidden",
            )
        return {
            "backend": backend,
            "language": "python",
            "native_humanize": True,
            "persistent_context": True,
        }
    if backend != "playwright":
        raise ConfigurationError(
            f"Unsupported Python automation backend: {backend}",
            code="automation_backend_unsupported",
        )
    return {
        "backend": backend,
        "language": "python",
        "framework_version": resolve_playwright_version(framework_version),
        "native_humanize": True,
        "persistent_context": True,
    }

File: packages/node/src/automation.ts
import { createRequire } from "node:module";

import { ConfigurationError } from "./errors.js";

export type AutomationBackend = "project-webdriver" | "playwright" | "puppeteer";
export type FrameworkBackend = Exclude<AutomationBackend, "project-webdriver">;

export interface AutomationCapability {
  backend: AutomationBackend;
  language: "node";
  frameworkVersion?: string;
  nativeHumanize: boolean;
  persistentContext: boolean;
}

const SUPPORTED_LINES: Readonly<Record<FrameworkBackend, readonly string[]>> = Object.freeze({
  playwright: Object.freeze(["1.62"]),
  puppeteer: Object.freeze(["25"]),
});

const PACKAGE_CANDIDATES: Readonly<Record<FrameworkBackend, readonly string[]>> = Object.freeze({
  playwright: Object.freeze(["playwright-core", "playwright"]),
  puppeteer: Object.freeze(["puppeteer-core", "puppeteer"]),
});

function versionLine(version: string): string {
  const match = version.match(/^(\d+)\.(\d+)(?:\.\d+)?(?:[-+][0-9A-Za-z.-]+)?$/);
  if (!match) throw new ConfigurationError(`Invalid framework version: ${version}`, "framework_version_invalid");
  return match[1] === "1" ? `${match[1]}.${match[2]}` : match[1]!;
}

export function validateFrameworkVersion(backend: FrameworkBackend, version: string): string {
  const line = versionLine(version);
  if (!SUPPORTED_LINES[backend].includes(line)) {
    throw new ConfigurationError(
      `Unsupported ${backend} version ${version}; supported lines: ${SUPPORTED_LINES[backend].join(", ")}`,
      "framework_version_unsupported",
    );
  }
  return version;
}

export function installedFrameworkVersion(backend: FrameworkBackend): string {
  const require = createRequire(import.meta.url);
  for (const packageName of PACKAGE_CANDIDATES[backend]) {
    try {
      const metadata = require(`${packageName}/package.json`) as { version?: unknown };
      if (typeof metadata.version === "string") return validateFrameworkVersion(backend, metadata.version);
    } catch (error) {
      const code = (error as NodeJS.ErrnoException).code;
      if (code !== "MODULE_NOT_FOUND" && code !== "ERR_PACKAGE_PATH_NOT_EXPORTED") throw error;
    }
  }
  throw new ConfigurationError(
    `Unable to determine the installed ${backend} package version; pass frameworkVersion explicitly`,
    "framework_version_missing",
  );
}

export function resolveFrameworkVersion(backend: FrameworkBackend, explicit?: string): string {
  return explicit === undefined
    ? installedFrameworkVersion(backend)
    : validateFrameworkVersion(backend, explicit);
}

export function automationCapability(
  backend: AutomationBackend = "project-webdriver",
  frameworkVersion?: string,
): AutomationCapability {
  if (backend === "project-webdriver") {
    if (frameworkVersion !== undefined) {
      throw new ConfigurationError("Project WebDriver does not accept frameworkVersion", "framework_version_forbidden");
    }
    return {
      backend,
      language: "node",
      nativeHumanize: true,
      persistentContext: true,
    };
  }
  return {
    backend,
    language: "node",
    frameworkVersion: resolveFrameworkVersion(backend, frameworkVersion),
    nativeHumanize: true,
    persistentContext: true,
  };
}

export function requireFrameworkHumanizeSupport(backend: FrameworkBackend, requested: boolean | undefined): void {
  void backend;
  void requested;
}
